### PR TITLE
Updated NBTTagString.toString() to escape '\' (and removed indexes from the output of NBTTagList.toString)

### PR DESCRIPTION
**Pull request by Mathe172**
This change ensures that NBTTagString.toString() generates  valid NBT. Without this, the output of commands like BlockData is invalid. Currently, quotation marks are escaped, but not backslashes, which is extremely inconsistent

EDIT: As explained in [this comment](https://github.com/MinecraftForge/MinecraftForge/pull/2393#issuecomment-177631438), this also fixes some edge cases which are broken in vanilla

### CODE REVIEW

**Comment by Mathe172** [comment]
Maybe to provide an example as to why this is necessary: The mod we're working on will allow you to store arbitrary NBTs in Strings, which can later be reinterpreted as NBTs. This reinterpretation breaks without this fix. The alternative would be to make an external toString converter for NBTs that handles toString of compounds and lists correctly (by calling custom toString methods for their items). But given the fact that the fix is a oneliner (and the current state is more or less broken), this PR seems like the better choice.

**Comment by asiekierka** [comment]
Wouldn't NBT<->JSON conversion be a better idea for your mod? Vanilla
already does it...

2016-01-27 12:48 GMT+01:00 Mathe172 <notifications@github.com>:

**Comment by Mathe172** [comment]
In theory, this would be possible, yes. But why would you ever use NBT->JSON->String if vanilla provides us with a NBT->String (although broken)? There is also another problem (and this is not a matter of "it would be nice if..."): Consider a String inside an NBT ending with a `\`. This String would return something broken when calling `toString`.
But considering that you do not have to escape `\` inside NBTs in vanilla (although you can, and you have to if they're followed by `"`) the problem might be bigger than just having a broken `toString` method...

**Comment by asiekierka** [comment]
Why would you use a hackjob over a true, portable object notation format,
especially if vanilla uses NBT->JSON->String itself?

YOU'RE NOT SUPPOSED TO USE TOSTRING AS OBJECT CONVERSION. It's like using a
nickname in place of a Social Security Number.

2016-01-31 15:03 GMT+01:00 Mathe172 <notifications@github.com>:

**Comment by Mathe172** [comment]
I don't see anything hacky in fixing broken code...
Btw, where is your NBT->Json used in vanilla? I looked through all files containing 'json' and 'nbt', but none of them contained anything close to a conversion from NBT to Json (the only thing converting NBT to something close to the json-format is `NBTBase.toString`, which is broken). We want to be able to do the reverse of String->NBT (as done by `JsonToNBT`, which doesn't even parse real Json), meaning that we need a readable and valid representation on NBTs (which is nearly done by `NBTBase.toString`).
If you could point us to the NBT->Json conversion, we could take a look at it, but somehow that doesn't seem to be possible, since Json doesn't have a concept of floats and doubles, everything is just a number (looking at http://www.json.org/).

**Comment by Lunatrius** [comment]
If you take a look at `CommandGive`, you'll notice `JsonToNBT.getTagFromJson(s)`. It's a subset of JSON and supports `73b`, `3847s`, `388.2333` etc.

**Comment by Mathe172** [comment]
This is the String->NBT conversion I was referring to (I have no idea why it is called `getTagFromJson`, since it has nothing to do with real Json as you said yourself). However, the reverse of this seems to be `NBTBase.toString` (although @asiekierka meant there's something else) and this is broken. This PR simply fixes this NBT->String conversion

**Comment by iTitus** [comment]
This might be out of context, but the NBTTagList does not have a toString method that produces valid JSON either.
It encapsulates the whole List in square brackets ('[' and ']'). This follows the JSON standard for arrays. But then it adds a comma separated list of `index:value` pairs. This is not valid JSON (for reference: [Wikipedia](https://en.wikipedia.org/wiki/JSON)).

Example of an enchanted item:
It uses a NBTTagList to save its enchantments.
This is the (JSON-like formatted) output of the toString method:
```
{
    ench: [
        0: {
            lvl: (level as short),
            id: (id as short)
        },
        1: {
            lvl: (level as short),
            id: (id as short)
        },
        ...
    ]
}
```

Valid JSON would be to just remove the `index:` thing or to encapsulate the whole pair in curly brackets to make it JSON object.

**Comment by darkhax** [comment]
I don't think it's safe to mess with the vanilla behavior in this situation. The valid way to save a NBT object is through JsonToNBT. Have you fully explored the impact that your change would have on the vanilla code? By changing the format of the toString output to how you think it should work might introduce new bugs in the vanilla code which expect the existing format.

**Comment by iTitus** [comment]
I would recommend to add a new `toValidJSON()` method

**Comment by Spartan322** [comment]
@iTitus That is probably the best solution, it doesn't leave a possible Forge/Vanilla Compat issue and yet guarantees JSON conversion, I'd rather get behind a `toValidJson()` method

**Comment by Mathe172** [comment]
Adding `toValidJSON` would require a lot more patching.
@darkhax @iTitus @Spartan322: The change proposed has no impact on vanilla, assuming `JsonToNBT` is to be used as reference for the NBT-format (in fact, it would fix the current code). As mentioned above, the current code in `toString` violates (in the special case that the last character of a `NBTTagString` is a `\`) the NBT-format as "specified" in `JsonToNBT`, which would be fixed by this PR.
@iTitus: You're right, the indexes produced by `NBTTagList.toString` are not JSON-like. I'll update the PR to remove the indexes, which might however be a debatable step, since `JsonToNBT` accepts the format with index (they seem to be simply ignored)

tl;dr: The current implementation broken in some edge cases, this PR fixes it and does not break vanilla-compatibility.

**Comment by iTitus** [comment]
I think if MC uss a pseudo-JSON format for their NBT I/O. We should leave that there.
You are though: it would require a lot of patching to add a `toValidJSON` method. So maybe an extra method, created by Forge in a Forge class?

**Comment by Spartan322** [comment]
A type of NBT Forge hook would probably be fine, it wouldn't require more then around 4 lines of patching to the base class.

**Comment by diesieben07** [comment]
@iTitus NBT when saved to disk is a binary file (NBT = Named _Binary_ Tag). The JSON-like representation is only used when interacting with players through e.g. command blocks.

**Comment by Mathe172** [comment]
@iTitus: Why don't you want to fix `NBTBase.toString`? As stated above, it's broken and our fix doesn't break anything (assuming `JsonToNBT` is always used to interpret Strings representing NBTs)
@Spartan322: If you're referring to implementing a `toValidJson`, I don't see how you'd do that by only patching the base class. If I'm missing something, could you elaborate what you mean?

**Comment by Lunatrius** [comment]
There are plenty of uses of `NBTBase.toString`. One of them is in `ItemStack.getChatComponent`:
```java
ichatcomponent.getChatStyle().setChatHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_ITEM, new ChatComponentText(nbttagcompound.toString())));
```

**Comment by iTitus** [comment]
@diesieben07 I was just over-generalizing here. I was too lazy to look up all the use cases, like @Lunatrius did. I knew that this was used for the command stuff.

**Comment by Spartan322** [comment]
@Mathe172 If you create a static method in a Forge class which directly takes the string representation (in its vanilla state) of the NBT and strips and converts all problems with the NBT types into JSON and then have the BaseNBT call it from inside a function, it would allow a public representation of JSON NBTs without modifying vanilla behavior.

**Comment by Mathe172** [comment]
@iTitus, @Lunatrius: We checked every single candidate for a call to `NBTBase.toString`. As far as we could tell, all of them are either printed for the user or are converted back to NBT using `JsonToNBT`, meaning this PR doesn't break anything (but it fixes some edge cases, as mentioned above)

**Comment by Mathe172** [comment]
I added a commit removing the indexes in `NBTTagList.toString` - as mentioned in the commit description, this change might be debatable:
Cons:
* Doesn't fix anything
Pros:
* Output more readable (imho)
* The indexes might suggest that `[0:"a",2:"c"]` creates a list of length 3, while it does not
* They do not add any information, as they are ignored

I'll revert the commit should it be deemed inappropriate or someone finds a problem

**Comment by Mathe172** [comment]
@Spartan322: If I understand you correctly, you're suggesting to call `NBTBase.toString` and fix the output afterwards - while highly inefficient from a performance perspective, it also breaks in the exact edge cases this PR attempts to fix

**Comment by Spartan322** [comment]
@Mathe172  I'm not asking you to use `toString()`, what I am asking is to pass a string representation of the NBT. What would the point of a method being added if you were using `toString()` anyway?

Anyway, doing this wouldn't result in much, if any, loss in efficiency then what you are already doing.

**Comment by Mathe172** [comment]
@Spartan322: What do you mean by "a string representation of the NBT"? I'm not quite sure where you'd get that representation.

**Comment by Spartan322** [comment]
Am I wrong at thinking field_74751_a is a string that represents the NBT data?

**Comment by Mathe172** [comment]
@Lunatrius: If you're trying to find a difference between the two implementations:
Consider the following command: `/give @p minecraft:iron_pickaxe 1 0 {display:{Lore:["test\\"]}}`
Without the PR, this does the following: Gives you a pick with lore `test\`. The hover message in chat displays `Invalid item!`
With the PR, the hover message correctly displays the item with lore.

@Spartan322: But you'd still have to create a `toValidJson`-converter traversing the NBT-tree and checking each type and handling it appropriately (i.e. return `data` for `NBTTagString`, list all members for `NBTTagCompound`, e.t.c.). While you'd get around patching each class (since you're doing it from the outside), you're still rewriting each `toString` method.
Also, I think the example given above clearly shows that the current implementation is broken, which should justify patching it.

**Comment by Spartan322** [comment]
Maybe, but I honestly don't think we should be modifying `toString()` because toString is not really designed for that, and even if vanilla doesn't get affected, its name doesn't actually describe the true purpose behind it (though I guess docing it can fix that, but then toString is still a Java Object standard), but even then, there is the possibility of breaking systems which use toString (even in the case where few people do that and we are a less stable API in 1.8.8/9, the patching is still giving a rather good possibility of breaking, depending on the implementation)

**Comment by iTitus** [comment]
This is a hacky workaround for displaying the NBT of an itemstack as formatted JSON on the tooltip:
https://gist.github.com/iTitus/313d06f2327d4b92ad8d

**Comment by LexManos** [comment]
The proper route to go without hacking things together is to write the opposite of Mojang's JSONToNBT, a NBTToJson class and stick it in forge.util.
Trying to find all the cases where toString is called and making it valid json and then deal with all the vanilla and 3rd party things that rely on that format would be annoying.
toString is never garenteed to be a convertible format. So the output that NBT does right now is completely valid.

**Comment by Mathe172** [comment]
What would you suggest to do about the broken Chat-events (see my comment above)? Would you rather fix this by not calling `NBTBase.toString` but the new "`NBTToJson`"? I get that that's just a minor issue, but it would be nice to have it fixed. (And not to seem stubborn, but the "fixed" output of `toString` is always equivalent to the old output if the old one is not broken, so there should be no impact by changing it - but you can probably judge that better ;-))
(I'm only talking about `NBTTagString` - the change in `NBTTagList` should probably be removed to remain as compatible as possible (we'll change that later))

**Comment by iTitus** [comment]
As you can see from my gist, the serialization of the ByteArray does not work either

**Comment by Mathe172** [comment]
@iTitus: Interesting, never noticed that. However, since you can't create a `NBTTagByteArray` using `JsonToNBT`, it might be debatable what the "proper" representation of a byte array is... (and whether it should be changed)
Of course, this poses another question: How does any of the code serializing items/blocks/... work when `NBTTagByteArray`s are used? While this is rarely the case (and should probably never be done), there's at least one case in vanilla where an NBT is serialized and deserialized to Json (the one I mentioned above)

**Comment by Earthcomputer** [comment]
I have a class which does the opposite of JsonToNBT [here] (https://github.com/Earthcomputer/Easy-Editors/blob/master/src/main/java/net/earthcomputer/easyeditors/api/util/NBTToJson.java).
It tries to produce the shortest possible valid input for JsonToNBT from the NBTTagCompound it is given. This kind of thing isn't too hard to write, actually, so I don't think we should go around modifying base classes where they don't cause problems in vanilla

**Comment by Mathe172** [comment]
@EarthComputer: We know that it's possible to make a working version without modifying base classes (there's also already code similar to yours in this thread). The point of this PR is to**fix** the currently **broken** vanilla version (as stated several times) - if you look at the issue that referenced this PR (above your comment), there are aslo real world examples that suffer from this

**Comment by Earthcomputer** [comment]
I'm sorry, I was unclear with my comment. Some of the things modified by this pull request are unnecessary, such as removing the indexes of NBTTagLists. This doesn't cause a problem as far as I know. Or, on second thoughts, maybe it does - my tests were not thorough.

**Comment by Mathe172** [comment]
As mentioned before ([here](https://github.com/MinecraftForge/MinecraftForge/pull/2393#issuecomment-182555276)), the indexes for `NBTTagList` probably shouldn't be removed (as stated above, we'll remove that change if requested). As for being a `toString` method instead of a `toJSON` method- if you had read the posts in this thread you'd know that even vanilla uses this method as an inverse of `JsonToNBT`, which doesn't always work (because it's **broken** - sorry to repeat that over and over, but we always have to reanswer the same questions).

**Comment by Actuarius** [comment]
@williewillus added labels [Bug]
